File: cpp/Neetcode/Blind75/Backtracking/CombinationSum.cpp
class Solution
{
public:
    vector<vector<int>> combs;
    vector<vector<int>> combinationSum(vector<int> &nums, int target)
    {
        vector<int> curr;
        backtrack(nums, curr, target, 0);
        return combs;
    }

    void backtrack(vector<int> &nums, vector<int> &curr, int target, int idx)
    {
        if (target == 0)
        {
            combs.push_back(curr);
            return;
        }

        if (target < 0 || idx >= nums.size())
        {
            return;
        }

        curr.push_back(nums[idx]);
        backtrack(nums, curr, target - nums[idx], idx);
        curr.pop_back();
        backtrack(nums, curr, target, idx + 1);
    }

    void backtrackLoop(vector<int> &nums, vector<int> &curr, int target, int start)
    {
        if (target == 0)
        {
            combs.push_back(curr);
            return;
        }

        if (target < 0 || start >= nums.size())
        {
            return;
        }

        for (int i = start; i < nums.size(); i++)
        {
            curr.push_back(nums[i]);
            backtrack(nums, curr, target - nums[i], i);
            curr.pop_back();
        }
    }
};

File: Python/Grind75/CombinationSum.py
from typing import List

# https://leetcode.com/problems/combination-sum/description/


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        combs = []
        n = len(candidates)
        candidates = sorted(candidates)

        def backtrack(comb, remaining, index):
            if remaining == 0:
                combs.append(comb[:])
                return
            elif remaining < 0 or index == n:
                return

            candidate = candidates[index]
            comb.append(candidate)
            backtrack(comb, remaining - candidate, index)
            comb.pop()
            backtrack(comb, remaining, index + 1)

        backtrack([], target, 0)
        return combs


if __name__ == "__main__":
    solution = Solution()
    print(solution.combinationSum([2, 3, 6, 7], 7))
